File: solutions/haskell/grains/1/src/Grains.hs
module Grains (square, total) where

p :: Integer -> Integer
p n = 2 ^ (n - 1)

square :: Integer -> Maybe Integer
square n
  | n < 1 || n > 64 = Nothing
  | otherwise = Just (p n)

total :: Integer
total = foldr ((+) . p) 0 [1 .. 64]

File: solutions/fsharp/grains/1/Grains.fs
module Grains

let square (n: int) : Result<uint64, string> =
    match n with
    | n when n < 1 || n > 64 -> Error "square must be between 1 and 64"
    | n -> Ok(pown 2UL (n - 1))

let total: Result<uint64, string> =
    [ 0..64 ]
    |> Seq.sumBy (fun n -> square n |> Result.defaultValue 0UL)
    |> fun result -> Ok result // I don't know why this needs to be wrapped
